Data providers don't expect empty projects on GitHub

While calculating security ratings with the demo tool, data providers sometimes fail with the following exception. It happens when a repository is empty. It may be better to make the data providers or the demo tool aware of such cases. For example, the tool can run some smoke tests for a project before passing it to the data providers.

```
[!] Couldn't fetch data, something went wrong!
org.kohsuke.github.HttpException: {"message":"Git Repository is empty.","documentation_url":"https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository"}
	at org.kohsuke.github.GitHubClient.interpretApiError(GitHubClient.java:440)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:368)
	at org.kohsuke.github.GitHubPageIterator.fetch(GitHubPageIterator.java:144)
	at org.kohsuke.github.GitHubPageIterator.hasNext(GitHubPageIterator.java:91)
	at org.kohsuke.github.PagedIterator.fetch(PagedIterator.java:106)
	at org.kohsuke.github.PagedIterator.nextPageArray(PagedIterator.java:134)
	at org.kohsuke.github.PagedIterable.toArray(PagedIterable.java:79)
	at org.kohsuke.github.PagedIterable.toArray(PagedIterable.java:107)
	at org.kohsuke.github.PagedIterable.toList(PagedIterable.java:119)
	at com.sap.sgs.phosphor.fosstars.data.github.GitHubDataFetcher.commitsFor(GitHubDataFetcher.java:97)
	at com.sap.sgs.phosphor.fosstars.data.github.GitHubDataFetcher.commitsFor(GitHubDataFetcher.java:83)
	at com.sap.sgs.phosphor.fosstars.data.github.GitHubDataFetcher.commitsAfter(GitHubDataFetcher.java:150)
	at com.sap.sgs.phosphor.fosstars.data.github.UsesGithubForDevelopment.usesGithubForDevelopment(UsesGithubForDevelopment.java:108)
	at com.sap.sgs.phosphor.fosstars.data.github.UsesGithubForDevelopment.fetchValueFor(UsesGithubForDevelopment.java:94)
	at com.sap.sgs.phosphor.fosstars.data.github.CachedSingleFeatureGitHubDataProvider.fetchValuesFor(CachedSingleFeatureGitHubDataProvider.java:29)
	at com.sap.sgs.phosphor.fosstars.data.github.CachedSingleFeatureGitHubDataProvider.fetchValuesFor(CachedSingleFeatureGitHubDataProvider.java:16)
	at com.sap.sgs.phosphor.fosstars.data.AbstractCachingDataProvider.doUpdate(AbstractCachingDataProvider.java:64)
	at com.sap.sgs.phosphor.fosstars.data.AbstractCachingDataProvider.doUpdate(AbstractCachingDataProvider.java:21)
	at com.sap.sgs.phosphor.fosstars.data.AbstractDataProvider.update(AbstractDataProvider.java:51)
	at com.sap.sgs.phosphor.fosstars.data.AbstractDataProvider.update(AbstractDataProvider.java:17)
	at com.sap.sgs.phosphor.fosstars.tool.github.SingleSecurityRatingCalculator.calculateFor(SingleSecurityRatingCalculator.java:72)
	at com.sap.sgs.phosphor.fosstars.tool.github.SingleSecurityRatingCalculator.calculateFor(SingleSecurityRatingCalculator.java:42)
	at com.sap.sgs.phosphor.fosstars.tool.github.MultipleSecurityRatingsCalculator.calculateFor(MultipleSecurityRatingsCalculator.java:57)
	at com.sap.sgs.phosphor.fosstars.tool.github.MultipleSecurityRatingsCalculator.calculateFor(MultipleSecurityRatingsCalculator.java:75)
	at com.sap.sgs.phosphor.fosstars.tool.github.SecurityRatingCalculator.processConfig(SecurityRatingCalculator.java:246)
	at com.sap.sgs.phosphor.fosstars.tool.github.SecurityRatingCalculator.run(SecurityRatingCalculator.java:174)
	at com.sap.sgs.phosphor.fosstars.tool.github.SecurityRatingCalculator.main(SecurityRatingCalculator.java:93)
Caused by: java.io.IOException: Server returned HTTP response code: 409 for URL: https://api.github.com/repos/eclipse/adore/commits
	at sun.reflect.GeneratedConstructorAccessor131.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1944)
	at sun.net.www.protocol.http.HttpURLConnection$10.run(HttpURLConnection.java:1939)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.net.www.protocol.http.HttpURLConnection.getChainedException(HttpURLConnection.java:1938)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1508)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:263)
	at org.kohsuke.github.GitHubHttpUrlConnectionClient$HttpURLConnectionResponseInfo.bodyStream(GitHubHttpUrlConnectionClient.java:197)
	at org.kohsuke.github.GitHubResponse$ResponseInfo.getBodyAsString(GitHubResponse.java:313)
	at org.kohsuke.github.GitHubResponse.parseBody(GitHubResponse.java:80)
	at org.kohsuke.github.GitHubPageIterator.lambda$fetch$0(GitHubPageIterator.java:145)
	at org.kohsuke.github.GitHubClient.createResponse(GitHubClient.java:406)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:360)
	... 25 more
Caused by: java.io.IOException: Server returned HTTP response code: 409 for URL: https://api.github.com/repos/eclipse/adore/commits
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1894)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:347)
	at org.kohsuke.github.GitHubHttpUrlConnectionClient.getResponseInfo(GitHubHttpUrlConnectionClient.java:69)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:348)
	... 25 more
```

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java
@@ -6,20 +6,30 @@ import java.io.UncheckedIOException;
 import java.time.Instant;
 import java.time.temporal.ChronoUnit;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Comparator;
 import java.util.Date;
 import java.util.List;
 import java.util.Optional;
 import org.apache.commons.collections4.map.LRUMap;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.kohsuke.github.GHCommit;
+import org.kohsuke.github.GHFileNotFoundException;
 import org.kohsuke.github.GHRepository;
 import org.kohsuke.github.GitHub;
+import org.kohsuke.github.HttpException;
 
 /**
  * Helper class for GitHub data providers which pulls the data from a GitHub repository. Also, the
  * class caches the fetched data for a certain amount of time.
  */
 public class GitHubDataFetcher {
+
+  /**
+   * A logger.
+   */
+  private static final Logger LOGGER = LogManager.getLogger(GitHubDataFetcher.class);
 
   /**
    * The singleton instance.
@@ -94,7 +104,12 @@ public class GitHubDataFetcher {
    * @throws IOException occurred during REST call to GitHub API.
    */
   private List<GHCommit> commitsFor(GHRepository repository) throws IOException {
-    return repository.listCommits().toList();
+    try {
+      return repository.listCommits().toList();
+    } catch (HttpException e) {
+      LOGGER.error(String.format("Could not fetch commits from %s", repository.getUrl()), e);
+      return Collections.emptyList();
+    }
   }
 
   /**
@@ -113,7 +128,18 @@ public class GitHubDataFetcher {
     }
 
     GHRepository repository = github.getRepository(project.path());
+    if (repository == null) {
+      throw new IOException(String.format("Could not fetch repository %s (null)", project.url()));
+    }
+
+    try {
+      repository.getDirectoryContent("/");
+    } catch (GHFileNotFoundException e) {
+      throw new IOException(String.format("Could not fetch content of / in %s", project.url()));
+    }
+
     repositoryCache.put(project, repository, expiration());
+
     return repository;
   }
 

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesDependabot.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesDependabot.java
@@ -12,8 +12,11 @@ import java.util.Date;
 import java.util.List;
 import org.kohsuke.github.GHCommit;
 import org.kohsuke.github.GHContent;
+import org.kohsuke.github.GHFileNotFoundException;
 import org.kohsuke.github.GHRepository;
+import org.kohsuke.github.GHUser;
 import org.kohsuke.github.GitHub;
+import org.kohsuke.github.GitUser;
 
 /**
  * <p>This data provider checks if an open-source project on GitHub
@@ -106,6 +109,8 @@ public class UsesDependabot extends CachedSingleFeatureGitHubDataProvider {
       }
 
       return content.getSize() >= ACCEPTABLE_CONFIG_SIZE;
+    } catch (GHFileNotFoundException e) {
+      return false; // okay, the config doesn't exist
     } catch (IOException e) {
       logger.warn("Something went wrong!", e);
     }
@@ -129,11 +134,47 @@ public class UsesDependabot extends CachedSingleFeatureGitHubDataProvider {
             || hasDependabotCommits(fetcher.commitsAfter(date, project, github)));
   }
 
+  /**
+   * Checks if a commit was done by Dependabot.
+   *
+   * @param commit The commit to be checked.
+   * @return True if the commit was done by Dependabot, false otherwise.
+   * @throws IOException If something went wrong.
+   */
   private static boolean isDependabot(GHCommit commit) throws IOException {
+    if (isDependabot(commit.getAuthor()) || isDependabot(commit.getCommitter())) {
+      return true;
+    }
+
     GHCommit.ShortInfo info = commit.getCommitShortInfo();
-    return commit.getAuthor().getLogin().toLowerCase().contains(DEPENDABOT_PATTERN)
-        || commit.getCommitter().getLogin().toLowerCase().contains(DEPENDABOT_PATTERN)
-        || info.getAuthor().getName().toLowerCase().contains(DEPENDABOT_PATTERN)
-        || info.getCommitter().getName().toLowerCase().contains(DEPENDABOT_PATTERN);
+    if (info == null) {
+      return false;
+    }
+
+    return isDependabot(info.getAuthor()) || isDependabot(info.getCommitter());
+  }
+
+  /**
+   * Check if a GitHub user looks like Dependabot.
+   *
+   * @param user The user to be checked.
+   * @return True if the user looks like Dependabot, false otherwise.
+   */
+  private static boolean isDependabot(GHUser user) {
+    return user != null
+        && user.getLogin() != null
+        && user.getLogin().toLowerCase().contains(DEPENDABOT_PATTERN);
+  }
+
+  /**
+   * Check if a git user looks like Dependabot.
+   *
+   * @param user The user to be checked.
+   * @return True if the user looks like Dependabot, false otherwise.
+   */
+  private static boolean isDependabot(GitUser user) {
+    return user != null
+        && user.getName() != null
+        && user.getName().toLowerCase().contains(DEPENDABOT_PATTERN);
   }
 }

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java
@@ -4,7 +4,7 @@ import static com.sap.sgs.phosphor.fosstars.model.feature.oss.OssFeatures.USES_G
 
 import com.sap.sgs.phosphor.fosstars.model.Feature;
 import com.sap.sgs.phosphor.fosstars.model.Value;
-import com.sap.sgs.phosphor.fosstars.model.value.BooleanValue;
+import com.sap.sgs.phosphor.fosstars.model.feature.oss.OssFeatures;
 import com.sap.sgs.phosphor.fosstars.tool.github.GitHubProject;
 import java.io.IOException;
 import java.time.Instant;
@@ -34,6 +34,9 @@ public class UsesGithubForDevelopment extends CachedSingleFeatureGitHubDataProvi
    */
   private static final double RESOLVED_USERS_THRESHOLD = 0.9;
 
+  /**
+   * A list of checks to figure out if a project uses GitHub for development.
+   */
   private static final List<Check> CHECKS = Arrays.asList(
 
       // check if the repository doesn't have an explicit link to the original repository
@@ -75,7 +78,7 @@ public class UsesGithubForDevelopment extends CachedSingleFeatureGitHubDataProvi
   );
   
   /**
-   * Initializes a data providerepository.
+   * Initializes a data provider.
    *
    * @param github An interface to the GitHub API.
    */
@@ -98,13 +101,13 @@ public class UsesGithubForDevelopment extends CachedSingleFeatureGitHubDataProvi
    * Checks if a project uses GitHub as a main development platform.
    *
    * @param project The project to be checked.
-   * @return {@link BooleanValue} of the feature.
+   * @return A value of {@link OssFeatures#USES_GITHUB_FOR_DEVELOPMENT}.
    */
   private Value<Boolean> usesGithubForDevelopment(GitHubProject project) {
     try {
       GHRepository repository = gitHubDataFetcher().repositoryFor(project, github);
 
-      Date threeMonthsAgo = Date.from(Instant.now().minus(3, ChronoUnit.MONTHS));
+      Date threeMonthsAgo = Date.from(Instant.now().minus(90, ChronoUnit.DAYS));
       List<GHCommit> commits = gitHubDataFetcher().commitsAfter(threeMonthsAgo, project, github);
 
       return USES_GITHUB_FOR_DEVELOPMENT.value(
@@ -136,7 +139,7 @@ public class UsesGithubForDevelopment extends CachedSingleFeatureGitHubDataProvi
   }
 
   /**
-   * Checks if user names from a commit point to existing GitHub users.
+   * Checks if user names from a commit point to real GitHub users.
    *
    * @param commit The commit to be checked.
    * @return True if user names from the commit can be resolved to existing GitHub users,

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/MultipleSecurityRatingsCalculator.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/MultipleSecurityRatingsCalculator.java
@@ -20,6 +20,11 @@ class MultipleSecurityRatingsCalculator extends AbstractRatingCalculator {
   private GitHubProjectCache projectCache = GitHubProjectCache.empty();
 
   /**
+   * A filename where the cache of projects should be stored.
+   */
+  private String projectCacheFile;
+
+  /**
    * A list of projects for which a rating couldn't be calculated.
    */
   private final List<GitHubProject> failedProjects = new ArrayList<>();
@@ -42,6 +47,17 @@ class MultipleSecurityRatingsCalculator extends AbstractRatingCalculator {
    */
   MultipleSecurityRatingsCalculator set(GitHubProjectCache projectCache) {
     this.projectCache = Objects.requireNonNull(projectCache, "Oh no! Project cache can't be null!");
+    return this;
+  }
+
+  /**
+   * Sets a file where the cache of processed projects should be stored.
+   *
+   * @param filename The file.
+   * @return The same {@link MultipleSecurityRatingsCalculator}.
+   */
+  MultipleSecurityRatingsCalculator storeProjectCacheTo(String filename) {
+    projectCacheFile = filename;
     return this;
   }
 
@@ -70,15 +86,22 @@ class MultipleSecurityRatingsCalculator extends AbstractRatingCalculator {
   @Override
   public MultipleSecurityRatingsCalculator calculateFor(List<GitHubProject> projects) {
     failedProjects.clear();
+
     for (GitHubProject project : projects) {
       try {
         calculateFor(project);
+
+        if (projectCacheFile != null) {
+          logger.info("Storing the project cache to {}", projectCacheFile);
+          projectCache.store(projectCacheFile);
+        }
       } catch (Exception e) {
         logger.warn("Oh no! Could not calculate a rating for {}", project.url());
         logger.warn(e);
         failedProjects.add(project);
       }
     }
+
     return this;
   }
 

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/SecurityRatingCalculator.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/SecurityRatingCalculator.java
@@ -234,19 +234,16 @@ public class SecurityRatingCalculator {
     }
 
     String projectCacheFile = projectCacheFile(config);
-    GitHubProjectCache projectCache = loadProjectCache(projectCacheFile);
 
     LOGGER.info("Starting calculating ratings ...");
-    MultipleSecurityRatingsCalculator calculator
-        = new MultipleSecurityRatingsCalculator(github, nvd);
-    calculator.set(projectCache);
-    calculator.set(VALUE_CACHE);
-    calculator.set(callback);
-    calculator.token(githubToken);
-    calculator.calculateFor(projects);
-
-    LOGGER.info("Storing the project cache to {}", projectCacheFile);
-    projectCache.store(projectCacheFile);
+    MultipleSecurityRatingsCalculator calculator =
+        (MultipleSecurityRatingsCalculator) new MultipleSecurityRatingsCalculator(github, nvd)
+            .set(loadProjectCache(projectCacheFile))
+            .storeProjectCacheTo(projectCacheFile)
+            .set(VALUE_CACHE)
+            .set(callback)
+            .token(githubToken)
+            .calculateFor(projects);
 
     LOGGER.info("Okay, we've done calculating the ratings");
 

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/SingleSecurityRatingCalculator.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/SingleSecurityRatingCalculator.java
@@ -1,6 +1,7 @@
 package com.sap.sgs.phosphor.fosstars.tool.github;
 
 import com.sap.sgs.phosphor.fosstars.data.DataProvider;
+import com.sap.sgs.phosphor.fosstars.data.github.GitHubDataFetcher;
 import com.sap.sgs.phosphor.fosstars.data.github.HasCompanySupport;
 import com.sap.sgs.phosphor.fosstars.data.github.HasSecurityPolicy;
 import com.sap.sgs.phosphor.fosstars.data.github.HasSecurityTeam;
@@ -53,10 +54,19 @@ class SingleSecurityRatingCalculator extends AbstractRatingCalculator {
   public SingleSecurityRatingCalculator calculateFor(GitHubProject project) throws IOException {
     Objects.requireNonNull(project, "Oh no! Project can't be null!");
 
-    OssSecurityRating rating = RatingRepository.INSTANCE.rating(OssSecurityRating.class);
-
     logger.info("Project: {}", project.url());
+
+    try {
+      GitHubDataFetcher.instance().repositoryFor(project, github);
+    } catch (IOException e) {
+      logger.error("Looks like something is wrong with the project!", e);
+      logger.warn("Let's skip the project ...");
+      return this;
+    }
+
     logger.info("Let's get info about the project and calculate a security rating");
+
+    OssSecurityRating rating = RatingRepository.INSTANCE.rating(OssSecurityRating.class);
 
     ValueSet values = ValueHashSet.unknown(rating.allFeatures());
     for (DataProvider<GitHubProject> provider : dataProviders()) {

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/SingleSecurityRatingCalculator.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/tool/github/SingleSecurityRatingCalculator.java
@@ -12,9 +12,13 @@ import com.sap.sgs.phosphor.fosstars.data.github.NumberOfCommits;
 import com.sap.sgs.phosphor.fosstars.data.github.NumberOfContributors;
 import com.sap.sgs.phosphor.fosstars.data.github.NumberOfStars;
 import com.sap.sgs.phosphor.fosstars.data.github.NumberOfWatchers;
+import com.sap.sgs.phosphor.fosstars.data.github.PackageManagement;
+import com.sap.sgs.phosphor.fosstars.data.github.ProgrammingLanguages;
 import com.sap.sgs.phosphor.fosstars.data.github.ProjectStarted;
 import com.sap.sgs.phosphor.fosstars.data.github.ScansForVulnerableDependencies;
-import com.sap.sgs.phosphor.fosstars.data.github.SecurityReviewForProject;
+import com.sap.sgs.phosphor.fosstars.data.github.UsesDependabot;
+import com.sap.sgs.phosphor.fosstars.data.github.UsesGithubForDevelopment;
+import com.sap.sgs.phosphor.fosstars.data.github.UsesNoHttpTool;
 import com.sap.sgs.phosphor.fosstars.data.github.UsesSignedCommits;
 import com.sap.sgs.phosphor.fosstars.data.interactive.AskAboutSecurityTeam;
 import com.sap.sgs.phosphor.fosstars.data.interactive.AskAboutUnpatchedVulnerabilities;
@@ -104,12 +108,16 @@ class SingleSecurityRatingCalculator extends AbstractRatingCalculator {
         new HasSecurityTeam(github),
         new HasCompanySupport(github),
         new HasSecurityPolicy(github),
-        new SecurityReviewForProject(github),
         new InfoAboutVulnerabilities(github, nvd),
         new IsApache(github),
         new IsEclipse(github),
         new LgtmDataProvider(github),
         new UsesSignedCommits(github),
+        new UsesDependabot(github),
+        new ProgrammingLanguages(github),
+        new PackageManagement(github),
+        new UsesNoHttpTool(github),
+        new UsesGithubForDevelopment(github),
         new ScansForVulnerableDependencies(github),
 
         // currently interactive data provider have to be added to the end, see issue #133

--- a/src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/MultipleSecurityRatingsCalculatorTest.java
+++ b/src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/MultipleSecurityRatingsCalculatorTest.java
@@ -3,11 +3,13 @@ package com.sap.sgs.phosphor.fosstars.tool.github;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.when;
 
 import com.sap.sgs.phosphor.fosstars.data.NoUserCallback;
+import com.sap.sgs.phosphor.fosstars.data.github.GitHubDataFetcher;
 import com.sap.sgs.phosphor.fosstars.model.math.DoubleInterval;
 import com.sap.sgs.phosphor.fosstars.model.rating.oss.OssSecurityRating.SecurityLabel;
 import com.sap.sgs.phosphor.fosstars.model.value.RatingValue;
@@ -16,14 +18,27 @@ import java.io.IOException;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import org.junit.After;
+import org.junit.Before;
 import org.junit.Test;
+import org.kohsuke.github.GHRepository;
 import org.kohsuke.github.GitHub;
 
 public class MultipleSecurityRatingsCalculatorTest {
 
+  @Before
+  @After
+  public void cleanup() {
+    GitHubDataFetcher.instance().repositoryCache().clear();
+    GitHubDataFetcher.instance().commitsCache().clear();
+  }
+
   @Test
-  public void calculateFor() throws IOException {
+  public void testCalculateFor() throws IOException {
     GitHub github = mock(GitHub.class);
+
+    GHRepository repository = mock(GHRepository.class);
+    when(github.getRepository(any())).thenReturn(repository);
 
     NVD nvd = new NVD();
 

--- a/src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/SingleSecurityRatingCalculatorTest.java
+++ b/src/test/java/com/sap/sgs/phosphor/fosstars/tool/github/SingleSecurityRatingCalculatorTest.java
@@ -3,11 +3,13 @@ package com.sap.sgs.phosphor.fosstars.tool.github;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.when;
 
 import com.sap.sgs.phosphor.fosstars.data.NoUserCallback;
+import com.sap.sgs.phosphor.fosstars.data.github.GitHubDataFetcher;
 import com.sap.sgs.phosphor.fosstars.data.github.IsApache;
 import com.sap.sgs.phosphor.fosstars.data.github.IsEclipse;
 import com.sap.sgs.phosphor.fosstars.model.math.DoubleInterval;
@@ -16,14 +18,27 @@ import com.sap.sgs.phosphor.fosstars.model.value.RatingValue;
 import com.sap.sgs.phosphor.fosstars.nvd.NVD;
 import java.io.IOException;
 import java.util.Arrays;
+import org.junit.After;
+import org.junit.Before;
 import org.junit.Test;
+import org.kohsuke.github.GHRepository;
 import org.kohsuke.github.GitHub;
 
 public class SingleSecurityRatingCalculatorTest {
 
+  @Before
+  @After
+  public void cleanup() {
+    GitHubDataFetcher.instance().repositoryCache().clear();
+    GitHubDataFetcher.instance().commitsCache().clear();
+  }
+
   @Test
-  public void calculateFor() throws IOException {
+  public void testCalculateFor() throws IOException {
     GitHub github = mock(GitHub.class);
+
+    GHRepository repository = mock(GHRepository.class);
+    when(github.getRepository(any())).thenReturn(repository);
 
     SingleSecurityRatingCalculator calculator
         = new SingleSecurityRatingCalculator(github, new NVD());
@@ -44,5 +59,46 @@ public class SingleSecurityRatingCalculatorTest {
     RatingValue ratingValue = apacheNiFi.ratingValue().get();
     assertEquals(SecurityLabel.BAD, ratingValue.label());
     assertTrue(DoubleInterval.closed(0, 3).contains(ratingValue.scoreValue().get()));
+  }
+
+  @Test
+  public void testCalculateForWithNoContent() throws IOException {
+    GitHub github = mock(GitHub.class);
+    GHRepository repository = mock(GHRepository.class);
+    when(github.getRepository(any())).thenReturn(repository);
+    when(repository.getDirectoryContent(any())).thenThrow(new IOException());
+    checkNoRating(github);
+  }
+
+  @Test
+  public void testCalculateForWithNullRepository() throws IOException {
+    GitHub github = mock(GitHub.class);
+    when(github.getRepository(any())).thenReturn(null);
+    checkNoRating(github);
+  }
+
+  @Test
+  public void testCalculateForWithExceptionWhenFetchingRepository() throws IOException {
+    GitHub github = mock(GitHub.class);
+    when(github.getRepository(any())).thenThrow(new IOException());
+    checkNoRating(github);
+  }
+
+  private static void checkNoRating(GitHub github) throws IOException {
+    SingleSecurityRatingCalculator calculator
+        = new SingleSecurityRatingCalculator(github, new NVD());
+
+    calculator.token("test");
+    calculator.set(NoUserCallback.INSTANCE);
+    calculator = spy(calculator);
+
+    when(calculator.dataProviders())
+        .thenReturn(Arrays.asList(new IsApache(github), new IsEclipse(github)));
+
+    GitHubProject apacheNiFi = new GitHubProject("apache", "nifi");
+    assertFalse(apacheNiFi.ratingValue().isPresent());
+
+    calculator.calculateFor(apacheNiFi);
+    assertFalse(apacheNiFi.ratingValue().isPresent());
   }
 }

